### PR TITLE
Add backup module for proxmox

### PR DESCRIPTION
**Pull request by IamLunchbox**
##### SUMMARY
This module adds a task to create backup tasks in a proxmox virtual environment cluster.  

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
proxmox_backup

##### ADDITIONAL INFORMATION
Currently, not all [api options](https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/{node}/vzdump) are covered, since they are either difficult to implement or only available for the root-user of the instance used.

For example, the api offers an exclude-mode next to the include, pool and all-mode. But this would would require other dependent settings to be set at as well. It could be added though, but I thought this will probably not be used, when `include`-mode and node-restrictions are available.

As to root-limited options (e.g. `tmpdir` api-parameter): I opted reproducing the options available in the GUI of a proxmox cluster, the options which I not implemented are not visible there either. And since I suspect most organizations would rather use a dedicated user or api-key to perform backups instead of the master-root user, I opted for simplicity and left these options out entirely. Users, which rely soley on the GUI to schedule and perform backups won't even know, that these options exist.

Compared to other proxmox modules, this module adds several checks regarding available permissions, but I value backups as mission critical and wanted to ensure a reasonable user experience by failing gracefully, if basic conditions are not met.
But on the other hand, this adds some complexity to the module, so if this is overburdening for an ansible module, i can delete them as well.

Lastly, I have seen PR #8360, but I have two issues with it: 
1. It does not implement what is states to do. It uses the [cluster/backup-api](https://pve.proxmox.com/pve-docs/api-viewer/index.html#/cluster/backup) endpoint, which is only used to **schedule** backups, not start them. Since it does not do what i need (actually starting backups, not just planning them), I opted for an implementation using the correct api endpoint (`nodes/{node}/vzdump`).
2. The aforementioned endpoint needs high privileges to begin with: (`Check: ["perm","/",["Sys.Modify"]]`). Any user who would want to perform backups using #8360 would essentially need root-privileges, while proxmox has a quite detailed permission system, which allows user to have a dedicated VM.Backup permission. This fine grained control is quite interesting for CI/CD setups, where an api-key with low level permissions can be used, to create a backup before for example a patch is deployed.  

If #8360 is merged at some point, I suggest it to be renamed to proxmox_backup_schedule, since that is what it does. I know that the proxmox api documentation is kind of misleading regarding the relationship of those two endpoints. If need be, I can post images of the GUI usage and the corresponding api-calls to the proxmox backend.

### CODE REVIEW

File: plugins/modules/proxmox_backup.py [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = r'''
+---
+module: proxmox_backup
+author: "Raphael Grieger (@IamLunchbox) <r.grieger@hotmail.com>"
+short_description: Start a VM backup in Proxmox VE cluster
+version_added: 10.0.0
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:17]: 10.0.0 is already out, so you cannot add new content to it :)
```suggestion
version_added: 10.1.0
```

File: plugins/modules/proxmox_backup.py [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+description:
+  - Allows you to create backups of KVM and LXC guests in Proxmox VE cluster.
+  - Offers the GUI functionality of creating a single backup as well as using the run-now functionality from the cluster backup schedule.
+  - The mininum required privileges to use this module are VM.Backup and Datastore.AllocateSpace.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:21]:
```suggestion
  - The mininum required privileges to use this module are C(VM.Backup) and C(Datastore.AllocateSpace).
```

File: plugins/modules/proxmox_backup.py [hunk 3]
```diff
@@ -0,0 +1,632 @@
...
+  check_mode:
+    support: full
+  diff_mode:
+    support: none
+  action_group:
+    version_added: 10.0.0
+options:
+  backup_mode:
+    description:
+      - The mode how proxmox performs backups. The default is, to create a runtime snapshot including memory.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:33]: The product Proxmox is written with a capital P. Please check other places in the documentation to make sure it's always spelled correctly.
```suggestion
      - The mode how Proxmox performs backups. The default is, to create a runtime snapshot including memory.
```

File: plugins/modules/proxmox_backup.py [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+      - Check U(https://pve.proxmox.com/pve-docs/chapter-vzdump.html#_backup_modes) for an explanation of the differences.
+    type: str
+    choices: ['snapshot', 'suspend', 'stop']
+    default: snapshot
+  bandwidth:
+    description:
+      - Limit the I/O bandwidth (in KiB/s) to send the backup. 0 is unlimited.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:40]:
```suggestion
      - Limit the I/O bandwidth (in KiB/s) to send the backup. V(0) is unlimited.
```

File: plugins/modules/proxmox_backup.py [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+    type: int
+  compress:
+    description:
+      - Enable additional compression of the backup archive.
+      - 0 will use the proxmox recommended value, depending on your storage target.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:45]:
```suggestion
      - V(0) will use the proxmox recommended value, depending on your storage target.
```

File: plugins/modules/proxmox_backup.py [hunk 6]
```diff
@@ -0,0 +1,632 @@
...
+    type: str
+    choices: [ '0', '1', 'gzip', 'lzo', 'zstd' ]
+  compression_threads:
+    description:
+      - The number of threads zstd will use to compress the backup.
+      - 0 uses 50% of the available cores, anything larger then 0 will use exactly as many threads.
+      - Is ignored if you specify O(compress=gzip) or O(compress=lzo)
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:52]:
```suggestion
      - V(0) uses 50% of the available cores, anything larger then V(0) will use exactly as many threads.
      - Is ignored if you specify O(compress=gzip) or O(compress=lzo).
```

File: plugins/modules/proxmox_backup.py [hunk 7]
```diff
@@ -0,0 +1,632 @@
...
+    type: int
+  description:
+    description:
+      - Specify the description for the backup.
+      - Template string for generating notes for the backup(s).
+      - Can contain variables which will be replaced by their values.
+      - Currently supported are {{cluster}}, {{guestname}}, {{node}}, and {{vmid}}, but more might be added in the future.
+      - Needs to be a single line, newline and backslash need to be escaped as `\n` and `\\` respectively.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:60]:
```suggestion
      - Currently supported are V({{cluster}}), V({{guestname}}), V({{node}}), and V({{vmid}}), but more might be added in the future.
      - Needs to be a single line, newline and backslash need to be escaped as V(\\n) and V(\\\\) respectively.
```

File: plugins/modules/proxmox_backup.py [hunk 8]
```diff
@@ -0,0 +1,632 @@
...
+    description:
+      - Specifices the mode to select backup targets.
+    choices: ['include', 'all', 'pool']
+    required: true
+    type: str
+  node:
+    description:
+      - Only execute the backup job for the given node.
+      - This option is usually used if O(mode=all).
+      - If you specify a node id and your vmids or pool do not reside there, they will not be backed up!
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:78]:
```suggestion
      - If you specify a node ID and your vmids or pool do not reside there, they will not be backed up!
```

File: plugins/modules/proxmox_backup.py [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+    description:
+      - Determine which notification system to use.
+    type: str
+    choices: ['auto','legacy-sendmail', 'notification-system']
+    default: auto
+  performance_tweaks:
+    description:
+      - Enable other performance-related settings.
+      - Must be entered as a string, containing comma separated key-value pairs.
+      - "For example: max-workers=2,pbs-entries-max=2."
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:90]:
```suggestion
      - "For example: V(max-workers=2,pbs-entries-max=2)."
```

File: plugins/modules/proxmox_backup.py [hunk 10]
```diff
@@ -0,0 +1,632 @@
...
+      - Marks backups as protected.
+      - >
+        "Might fail, when the PBS backend has verify enabled
+         due to this bug: U(https://bugzilla.proxmox.com/show_bug.cgi?id=4289)"
+    type: bool
+  retention:
+    description:
+      - >
+        Use custom retention options instead of those from the default cluster
+        configuration (which is usually 'keep-all').
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:109]:
```suggestion
        configuration (which is usually V(keep-all)).
```

File: plugins/modules/proxmox_backup.py [hunk 11]
```diff
@@ -0,0 +1,632 @@
...
+      - Always requires Datastore.Allocate permission at the storage endpoint.
+      - >
+        Specifying a retention time other than V(keep-all=1) might trigger pruning on the datastore,
+        if an existing backup should be deleted target due to your specified timeframe.
+      - Deleting requires Datastore.Modify or Datastore.Prune permissions on the backup storage.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:114]:
```suggestion
      - Deleting requires C(Datastore.Modify) or C(Datastore.Prune) permissions on the backup storage.
```

File: plugins/modules/proxmox_backup.py [hunk 12]
```diff
@@ -0,0 +1,632 @@
...
+    type: bool
+    default: false
+  wait_timeout:
+    description:
+      - Seconds to wait for the backup to be finished.
+      - Will only be evaluated, if O(wait=true).
+    type: int
+    default: 10
+notes:
+  - Requires proxmoxer and requests modules on host. These modules can be installed with pip.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:139]:
```suggestion
```
That's already said implicitly through `requirements`, I don't think it needs to be repeated.

File: plugins/modules/proxmox_backup.py [hunk 13]
```diff
@@ -0,0 +1,632 @@
...
+
+RETURN = r'''
+backups:
+  description: List of nodes and their task IDs.
+  returned: on success
+  type: list
+  elements: dict
+  contains:
+    node:
+      description: Node id.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:206]:
```suggestion
      description: Node ID.
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments on the documentation.

File: plugins/modules/proxmox_backup.py (hunk 1 above)
Comment (IamLunchbox) [reply plugins/modules/proxmox_backup.py:17]: time flies...

**Comment by felixfontein** [comment]
`version_added` for attributes is only needed if they are added in later versions. So you can remove it.

Also note that you did not add the module to the action group. Only the modules listed for the `proxmox` action group in `meta/runtime.yml` are actually part of it.

**Comment by IamLunchbox** [comment]
Ok, thanks.  I will implement those two fixes today.

File: plugins/modules/proxmox_backup.py [hunk 14]
```diff
@@ -0,0 +1,634 @@
...
+---
+module: proxmox_backup
+author: "Raphael Grieger (@IamLunchbox) <r.grieger@hotmail.com>"
+short_description: Start a VM backup in Proxmox VE cluster
+version_added: 10.1.0
+description:
+  - Allows you to create backups of KVM and LXC guests in Proxmox VE cluster.
+  - Offers the GUI functionality of creating a single backup as well as using the run-now functionality from the cluster backup schedule.
+  - The mininum required privileges to use this module are C(VM.Backup) and C(Datastore.AllocateSpace).
+  - Most options are optional and if unspecified will be chosen by the Cluster and its default values.
```
Comment (felixfontein) [inline plugins/modules/proxmox_backup.py:22]:
```suggestion
  - Most options are optional and if unspecified will be chosen by the Cluster and its default values.
  - Note that this module B(is not idempotent). It always starts a new backup (when not in check mode).
```

**Comment by felixfontein** [comment]
If nobody objects, I'll merge this PR for the next release.

File: plugins/modules/proxmox_backup.py (hunk 14 above)
Comment (IamLunchbox) [reply plugins/modules/proxmox_backup.py:22]: Added this note. Thanks, good point.

**Comment by felixfontein** [comment]
@IamLunchbox thanks for your contribution!

**Pull request merged**